### PR TITLE
[IMP] *: Graph Area Style + Fill Temporal Groups

### PR DESCRIPTION
**Pull request by FVolral**
Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: odoo/addons/base/tests/test_misc.py [hunk 1]
```diff
@@ -40,6 +42,115 @@ def test_repeated(self):
...
+            datetime.datetime(1985, 9, 1, 0, 0),
+            datetime.datetime(1985, 10, 1, 0, 0),
+            datetime.datetime(1985, 11, 1, 0, 0),
+            datetime.datetime(1985, 12, 1, 0, 0),
+            datetime.datetime(1986, 1, 1, 0, 0)
+        ]
+
+        dates = [date for date in date_utils.date_range(start_date, end_date)]
+
+        assert dates == expected
```
Comment (rco-odoo) [inline odoo/addons/base/tests/test_misc.py]: use assert methods
```python
self.assertEqual(dates, expected)
```

File: odoo/addons/base/tests/test_misc.py [hunk 2]
```diff
@@ -40,6 +42,98 @@ def test_repeated(self):
...
+            (datetime.datetime(1985, 7, 1, 0, 0), True),
+            (datetime.datetime(1985, 8, 1, 0, 0), True),
+            (datetime.datetime(1985, 9, 1, 0, 0), True),
+            (datetime.datetime(1985, 10, 1, 0, 0), False),
+            (datetime.datetime(1985, 11, 1, 0, 0), False),
+            (datetime.datetime(1985, 12, 1, 0, 0), False),
+            (datetime.datetime(1986, 1, 1, 0, 0), False),
+        ]
+
+        expected = [timezone.localize(dt, is_dst=is_dst) for dt, is_dst in expected]
```
Comment (rco-odoo) [inline odoo/addons/base/tests/test_misc.py]: Is the flag `is_dst` necessary?
In my opinion, it should not.

File: odoo/addons/base/tests/test_misc.py [hunk 3]
```diff
@@ -40,6 +42,98 @@ def test_repeated(self):
...
+
+    def test_date_range_with_hour(self):
+        """ Test date range with hour and naive timezone datetime."""
+        start_date = datetime.datetime(1985, 1, 1)
+        end_date = datetime.datetime(1985, 1, 2)
+        step_dt = relativedelta(hours=1)
+
+        dates = [date for date in date_utils.date_range(start_date, end_date, step_dt)]
+
+        self.assertEqual(len(dates), 25)
```
Comment (rco-odoo) [inline odoo/addons/base/tests/test_misc.py]: make the expected result explicit

File: addons/web/static/src/js/core/rpc.js [hunk 4]
```diff
@@ -67,6 +67,7 @@ return {
             var orderBy = options.orderBy || params.orderBy || params.kwargs.orderby;
             params.kwargs.orderby = orderBy ? this._serializeSort(orderBy) : orderBy;
             params.kwargs.lazy = 'lazy' in options ? options.lazy : params.lazy;
+            params.kwargs.fill_temporal = options.fill_temporal;
```
Comment (ged-odoo) [inline addons/web/static/src/js/core/rpc.js]: remove this

File: addons/web/static/src/js/views/graph/graph_renderer.js [hunk 5]
```diff
@@ -264,30 +278,37 @@ return AbstractRenderer.extend({
      * @returns {nvd3 chart}
      */
     _renderLineChart: function () {
-        if (this.state.data.length < 2) {
+        var self = this;
+
+        // Remove Undefined
```
Comment (ged-odoo) [inline addons/web/static/src/js/views/graph/graph_renderer.js]: did you try to remove undefined?

File: addons/web/static/tests/core/rpc_tests.js [hunk 6]
```diff
@@ -182,6 +182,7 @@ QUnit.module('core', {}, function () {
             groupBy: ['product_id'],
             context: {abc: 'def'},
             lazy: true,
+            fill_temporal: true,
```
Comment (ged-odoo) [inline addons/web/static/tests/core/rpc_tests.js]: remove

File: addons/web/static/tests/core/rpc_tests.js [hunk 7]
```diff
@@ -192,6 +193,7 @@ QUnit.module('core', {}, function () {
                 fields: ['name'],
                 groupby: ['product_id'],
                 lazy: true,
+                fill_temporal: true,
```
Comment (ged-odoo) [inline addons/web/static/tests/core/rpc_tests.js]: remove

File: odoo/models.py [hunk 8]
```diff
@@ -1746,6 +1747,108 @@ def _read_group_fill_results(self, domain, groupby, remaining_groupbys,
...
+            # in some circumstances, dates can have a different timezone resulting
+            # in a false negative. To avoid that, we remove that timezone but it
+            # could be an error.
+            for dt in expected:
+                if remove_tz:
+                    data[0][groupby_name] = data[0][groupby_name].replace(tzinfo=None)
+
+                if data[0][groupby_name] == dt:
+                    new_data.append(data.popleft())
+                else:
```
Comment (odony) [inline odoo/models.py]: we can simplify and preserve NULLS ordering by modifying as follows, without needing `nulls` etc:

```
if not data[0][groupby_name] or data[0][groupby_name] == dt:
    new_data.append(data.popleft())
else:
    new_data.append(dict(empty_data, **{groupby_name: dt}))
```

File: odoo/models.py [hunk 9]
```diff
@@ -1746,6 +1747,108 @@ def _read_group_fill_results(self, domain, groupby, remaining_groupbys,
...
+                nulls = []
+                data = collections.deque(data)
+
+            new_data = []
+
+            # Note: the list 'expected' contains the the dates that should be
+            # represented inside data when it is returned. Notice that 'expected'
+            # is sorted like data, so what we do is empty data progressivly by
+            # popping its elements from first to last. To do this properly, we
+            # need to compare dates together, but here comes a little problem :
```
Comment (odony) [inline odoo/models.py]: remove the part about 'here comes a little...'

File: odoo/tools/date_utils.py [hunk 10]
```diff
@@ -178,3 +177,45 @@ def json_default(obj):
...
         return fields.Date.to_string(obj)
     return ustr(obj)
+
+def date_range(start, end, step=relativedelta(months=1)):
+    """Date range generator with a step interval.
+
+    :param start: A datetime. begining date of the range.
+    :param end: A datetime. ending date of the range.
+    :param step: A relativedelta. interval of the range.
+    :return: A Generator[datetime, None, None].
```
Comment (odony) [inline odoo/tools/date_utils.py]: update docstring, use :rtype:  and the  :param type var: syntax

File: odoo/models.py [hunk 11]
```diff
@@ -1746,6 +1747,108 @@ def _read_group_fill_results(self, domain, groupby, remaining_groupbys,
...
+        first, last = existing[0], existing[-1]
+
+        # When the date_trunc function is used on a DATE (not a TIMESTAMP),
+        # PostGreSQL implicitly cast it as a TIMESTAMP WITH TIMEZONE. That time-
+        # zone is based on the local machine executing Postgres. We don't want
+        # this because it could cause various problems. That's why we detect
+        # those TZs by assuming they are always an instance of
+        # psycopg2.tz.FixedOffsetTimezone and remove them. Other possible
+        # timezone should come from the user context.
+        remove_tz = isinstance(first.tzinfo, psycopg2.tz.FixedOffsetTimezone)
```
Comment (odony) [inline odoo/models.py]: replace this hack by the proper casting to `::timestamp` [here](https://github.com/odoo/odoo/blob/e7b973b95326dd2f886d51537669b50fa09a08ba/odoo/models.py#L1942), so you can also remove other code dealing with `remove_tz` below

File: odoo/tools/date_utils.py [hunk 12]
```diff
@@ -178,3 +177,45 @@ def json_default(obj):
...
+            raise ValueError("Timezones of start argument and end argument seem inconsistent")
+
+    if not are_naive and not are_utc and not are_others:
+        raise ValueError("Timezones of start argument and end argument mismatch")
+
+    if start == end:
+        raise ValueError("start and end can't be equal")
+
+    if start > end:
+        start, end = end, start
```
Comment (odony) [inline odoo/tools/date_utils.py]: raise instead of auto-swapping boundaries

File: odoo/tools/date_utils.py [hunk 13]
```diff
@@ -178,3 +177,45 @@ def json_default(obj):
...
+
+    if are_others:
+        if start.tzinfo.zone != end.tzinfo.zone:
+            raise ValueError("Timezones of start argument and end argument seem inconsistent")
+
+    if not are_naive and not are_utc and not are_others:
+        raise ValueError("Timezones of start argument and end argument mismatch")
+
+    if start == end:
+        raise ValueError("start and end can't be equal")
```
Comment (odony) [inline odoo/tools/date_utils.py]: don't block this edge case, if the implementation works, it seems valid + fix corresponding test

**Pull request closed without merging**